app/src/main/java/com/google/android/sambadocumentsprovider/ShareManager.java

```diff
@@ -57,18 +60,37 @@
   private final Map<String, String> mServerStringMap = new HashMap<>();
   private final CredentialCache mCredentialCache;
 
+  private EncryptionManager mEncryptionManager;
+
   private final List<MountedShareChangeListener> mListeners = new ArrayList<>();
 
   ShareManager(Context context, CredentialCache credentialCache) {
     mCredentialCache = credentialCache;
 
+    try {
+      mEncryptionManager = new EncryptionManager(context);
+    } catch (EncryptionException e) {
+      Log.e(TAG, "Failed to init EncryptionManager: ", e);
+    }
+
     mPref = context.getSharedPreferences(SERVER_CACHE_PREF_KEY, Context.MODE_PRIVATE);
     // Loading saved servers.
     final Set<String> serverStringSet =
         mPref.getStringSet(SERVER_STRING_SET_KEY, Collections.<String> emptySet());
+
     final Map<String, ShareTuple> shareMap = new HashMap<>(serverStringSet.size());
+    final List<String> forceEncryption = new ArrayList<>();
     for (String serverString : serverStringSet) {
-      // TODO: Add decryption
+      try {
+        if (mEncryptionManager != null) {
```

Comment:
If we don't need to worry about the fallback, we can get rid of this null check.

You can use a stub implementation of EncryptionManager (abstract it to an interface) to avoid null check as well if you think fallback is necessary.